feat(xo-web/import/esxi): import VM from ESXi

rajaa212 opened the pull request:
- Backend PR: https://github.com/vatesfr/xen-orchestra/pull/6662
- List expected from the function to connect to ESXi:
```js
[
  {
    id: '1',
    nameLabel: 'test_linux',
    memory: 2147483648,
    nCpus: 1,
    guestToolsInstalled: false,
    firmware: 'bios',
    powerState: 'poweredOff',
    storage: {
      used: 216761,
      free: 2318061568,
    },
  },
  {
    id: '2',
    nameLabel: 'test_small',
    memory: 2147483648,
    nCpus: 1,
    guestToolsInstalled: false,
    firmware: 'bios',
    powerState: 'poweredOff',
    storage: {
      used: 10696,
      free: 2586497520,
    },
  },
]
```
- Params provided to `vm.importFromEsxi`
```js
 {
    host: '192.168.2.20',
    network: '6b309876-9b61-6feb',
    password: 'password',
    sr: '86a9757d-9c05-9fe0-e79a',
    sslVerify: false,
    user: 'user',
    vm: '1',
    },
  }
  
```

### Screenshots

![Capture d’écran du 2023-02-16 21-12-04](https://user-images.githubusercontent.com/7724491/219476856-79a24509-49cb-48a1-bf41-e5ba63c4fc5b.png)


![Capture d’écran du 2023-02-16 21-12-45](https://user-images.githubusercontent.com/7724491/219476851-6fd229ae-3edf-4a02-a9a7-1a4451cc76ce.png)


![Capture d’écran du 2023-02-21 16-08-11](https://user-images.githubusercontent.com/7724491/220383872-c0935e16-75e5-4afb-baaa-3357762ccb9d.png)


![Capture d’écran du 2023-02-21 16-13-15](https://user-images.githubusercontent.com/7724491/220383868-7c0184ed-7976-4c87-bac0-bbdecc73950f.png)



### Check list

> Check if done, if not relevant leave unchecked.

- [ ] PR reference the relevant issue (e.g. `Fixes #007` or `See xoa-support#42`)
- [x] if UI changes, a screenshot has been added to the PR
- [ ] documentation updated
- `CHANGELOG.unreleased.md`:
  - [x] enhancement/bug fix entry added
  - [x] list of packages to release updated (`${name} v${new version}`)
- **I have tested added/updated features** (and impacted code)
  - [ ] unit tests (e.g. [`cron/parse.spec.js`](https://github.com/vatesfr/xen-orchestra/blob/b24400b21de1ebafa1099c56bac1de5c988d9202/%40xen-orchestra/cron/src/parse.spec.js))
  - [ ] if `xo-server` API changes, the corresponding test has been added to/updated on [`xo-server-test`](https://github.com/vatesfr/xen-orchestra/tree/master/packages/xo-server-test)
  - [x] at least manual testing

### Process

1. create a PR as soon as possible
1. mark it as `WiP:` (Work in Progress) if not ready to be merged
1. when you want a review, add a reviewer (and only one)
1. if necessary, update your PR, and re- add a reviewer

From [_the Four Agreements_](https://en.wikipedia.org/wiki/Don_Miguel_Ruiz#The_Four_Agreements):

1. Be impeccable with your word.
1. Don't take anything personally.
1. Don't make assumptions.
1. Always do your best.

rajaa212 committed:
feat(xo-web/import/esxi): import VM from ESXI

rajaa212 committed:
support selecting ESXI

rajaa212 committed:
improved code to increase readability

rajaa212 committed:
improvement code

rajaa212 committed:
fixes and remove unnecessary changes

rajaa212 committed:
refactor duplicate code

rajaa212 committed:
handle when pool is undefined

fbeauchamp commented:
Tests done : 

* can't import from ova ( when dropping an ova it does not show the network/ disk form ) 
* can't import from xva import button stay disabled when dropping a xva 
* can't select ova format when importing from url 
* can't unselect ssl certificate page when importing to esxi ( thus, I am not able to test esxi import since our server need to uncheck this)

rajaa212 committed:
minor changes

olivierlambert commented:
I would prefer a dedicated page, as explained in the original issue (Import/From VMware), allowing multi VMs import, which is a special process we don't have in the "Import/VM" view.

rajaa212 committed:
handle ESXI import

rajaa212 committed:
fixes from feedbacks

rajaa212 committed:
fixes from feedbacks

rajaa212 committed:
more fixes & remove unnecessary changes

rajaa212 commented:
> Tests done :
> 
>     * can't import from ova ( when dropping an ova it does not show the network/ disk form )
> 
>     * can't import from xva import button stay disabled when dropping a xva
> 
>     * can't select ova format when importing from url
> 
>     * can't unselect ssl certificate page when importing to esxi ( thus, I am not able to test esxi import since our server need to uncheck this)

- Fixed
- Fixed
- It's already the case: URL import is only for XVA files
- Fixed

rajaa212 commented:
> I would prefer a dedicated page, as explained in the original issue (Import/From VMware), allowing multi VMs import, which is a special process we don't have in the "Import/VM" view.

- As a first step, we decided to allow the import of one VM.

fbeauchamp commented:
test done : rebase from master ( since there is fix on import, merge from branch https://github.com/vatesfr/xen-orchestra/pull/6662 ) 

* import ova /xva  : import ok, but got error : 
![image](https://user-images.githubusercontent.com/50174/219622914-513aa506-eb8f-421a-a93e-3b1d9fff71b3.png)
* import from esxi : 
  * connect ok
  * list VMs  show a list without any label 
  * the VM to import don't have any detail except memory 
  * clicking on the import give an invalid parameter message
[2023-02-17T10_59_59.772Z - XO.log](https://github.com/vatesfr/xen-orchestra/files/10765992/2023-02-17T10_59_59.772Z.-.XO.log)

rajaa212 committed:
fix ova/xva import

fbeauchamp commented on packages/xo-web/src/xo-app/vm-import/url-import.js:
```diff
@@ -0,0 +1,113 @@
+        const file = {
+          name: decodeURIComponent(url.slice(url.lastIndexOf('/') + 1)),
+        }
+        importVm(file, type.value, undefined, this.state.sr, url)
```
promises should be awaited

fbeauchamp commented on packages/xo-web/src/xo-app/vm-import/xva-import.js:
```diff
@@ -0,0 +1,202 @@
+      handleImport:
+        () =>
+        ({ sr, vms }) => {
+          importVms(
```
promise should be awaited

rajaa212 committed:
vmId instead vmData

fbeauchamp commented:
# Test done
## test Protocol
* pulled this branch
* rebase master onto it, since there is fixes on ova /vmdk handling
* merged the backend branch into it to get new response format from `esxi.connect` api method 

## import XVA 
* the `import` button does not stay disabled during the import
* a `succes` alert should be shown at the end
* user should be redirected to imported VM when done
* Not tested : the button should turn orange on import error
* import seems to have worked

## import Ova 
* the `import` button does not stay disabled with text `importing` during import
* user should be redirected to imported VM when done 
* a `succes` alert should be shown at the end
* Not tested : the button should turn orange on import error
* import seems to have worked

## import from url 
* not tested

## import from esxi 
* connect ok 
* connect with improper parameters : show error ok
* vm listing ok 
* network should have a default value ( as during the ova import ) 
* additionnal fields : not handled by backend or now 
* the `import` button does not stay disabled with text `importing` during import
* a `succes` alert should be shown at the end
* user should be redirected to imported VM when done (when importing 1 VM )

fbeauchamp committed:
Fix(@xen-orchestra/vmware-explorer): handle Esxi with more than 100 VMs

fbeauchamp committed:
feat(xo-server): expose method for Esxi import UX

rajaa212 committed:
remove unnecessary changes & code imrpoved

rajaa212 committed:
improve displayed VM config

fbeauchamp committed:
add missing file

rajaa212 committed:
changelog

fbeauchamp commented:
## Test
### Protocol
branch are up to date and includes necessary backend method 

### ova /xva import  
* got a translation problem : 
![image](https://user-images.githubusercontent.com/50174/220403543-53e8989e-c9a7-41cf-8f4c-42d46fe225a7.png)
* button : OK 
* import : ok
* success toast ok
* redirect after importing : Not ok

### from esxi 
* bad connect data : error message is ok
* connect with the right data : ok
* Vm listing : ok
* importing fail with 
![image](https://user-images.githubusercontent.com/50174/220404249-985f437e-340a-4151-a419-ef8001fe7c29.png)
Reselecting the network and it works
* importing : ok 
* button : ok

rajaa212 committed:
fixes

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/esxi-import.js:
```diff
@@ -0,0 +1,262 @@
+            user,
+            vm: vm.value,
+          })
+        },
```
You can simply return the promise instead of using async/await.

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/esxi-import.js:
```diff
@@ -0,0 +1,262 @@
+              handler={importVm}
+              icon='import'
+              redirectOnSuccess={
+                vmData === undefined ? undefined : `/home?s=name_label:${encodeURIComponent(vmData.nameLabel)}&t=VM`
```
Why don't you redirect to the VM page instead? Also, the name of the VM might not be enough to filter that one VM.

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/esxi-import.js:
```diff
@@ -0,0 +1,262 @@
+        () =>
+        async ({ hostIp, hasCertificate, password, user }) => {
+          const vms = await esxiConnect(hostIp, user, password, hasCertificate)
+          return { isConnected: true, vmsById: vms.reduce((vms, vm) => ({ ...vms, [vm.id]: vm }), {}) }
```
You can use `lodash`'s `keyBy` instead.

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/esxi-import.js:
```diff
@@ -0,0 +1,262 @@
+          return { isConnected: true, vmsById: vms.reduce((vms, vm) => ({ ...vms, [vm.id]: vm }), {}) }
+        },
+      linkState,
+      onChangeVm: (_, vm) => state => ({ vm, vmData: state.vmsById[vm.value] }),
```
Can't we just pass `vms` to the Select so that we don't need to have `vm` *and* `vmData`?

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/esxi-import.js:
```diff
@@ -0,0 +1,262 @@
+          </div>
+        </form>
+      )}
+      {isConnected && (
```
You could use a ternary operator instead of 2 opposite conditions. That way, if we change the condition or the name of the variable, we wouldn't have to change it in 2 places.

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/vm-data.js:
```diff
@@ -0,0 +1,57 @@
+    <Row>
+      <Col mediumSize={6}>
+        <div className='form-group'>
+          <label>{_('vmNameLabel')}:</label> <span>{data.nameLabel}</span>
```
The `:` should be part of the i18n because the space before it changes depending on the language. We have a `keyValue` message exactly for that case.

rajaa212 committed:
fixes from feedback

fbeauchamp commented:
## Test 

### protocol
only need to pull the branch

## xva / ova import : ok 

## import esxi : still running , but no error on start

fbeauchamp reviewed (approved):
functionnal tests are ok for me

rajaa212 committed:
remove unnecassary changes

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/esxi-import.js:
```diff
@@ -0,0 +1,260 @@
+    poolId => (poolId === undefined ? undefined : sr => isSrWritable(sr) && sr.$poolId === poolId)
+  )
+
+  _importVm = async () => {
```
You don't need to make it `async`.

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/esxi-import.js:
```diff
@@ -0,0 +1,260 @@
+
+  _toggleCertificateCheck = ({ target: { checked, name } }) => {
+    this.setState({ [name]: checked })
+  }
```
You can use `linkState`/`toggleState` to remove most of those.

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/esxi-import.js:
```diff
@@ -0,0 +1,260 @@
+
+    return (
+      <div>
+        {!isConnected ? (
```
Since it's almost the whole component, I think it would be more readable to write it like this:
```js
if (!isConnected) {
  return ...
}

return ...
```

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/esxi-import.js:
```diff
@@ -0,0 +1,260 @@
+              >
+                {_('newImport')}
+              </ActionButton>
+              <Button onClick={this._reset}>{_('formReset')}</Button>
```
We should probably also add a button to "disconnect" that would simply assign `false` to `isConnected`, what do you think?

pdonias commented on packages/xo-web/src/xo-app/vm-import/esxi/vm-data.js:
```diff
@@ -0,0 +1,66 @@
+    <Row>
+      <Col mediumSize={6}>
+        <div className='form-group'>
+          {_('keyValue', { key: intl.formatMessage(messages.vmNameLabel), value: data.nameLabel })}
```
Can't you pass `_('vmNameLabel')` here?

rajaa212 committed:
fixes from feedback

rajaa212 committed:
remove unnecassary changes

rajaa212 committed:
remove unnecassary changes

pdonias reviewed (approved):
LGTM but I don't know how to test it. @Rajaa-BARHTAOUI or @fbeauchamp, can you do a last end-to-end test after the latest changes?

rajaa212 commented:
> LGTM but I don't know how to test it. @Rajaa-BARHTAOUI or @fbeauchamp, can you do a last end-to-end test after the latest changes?

It has been tested after the latest changes, and everything is still running.

pdonias commented:
Ok, then this PR is ready to merge after #6662.

fbeauchamp committed:
fix(xo-web): update method to reflect changes in PR #6662

pdonias committed:
Merge branch 'master' into importEsxiUI
